filedescriptor/src/unix.rs

```diff
@@ -146,6 +166,39 @@ impl OwnedHandle {
         }
     }
 
+    #[inline]
+    pub(crate) unsafe fn dup2_impl<F: AsRawFileDescriptor>(
+        fd: &F,
+        dest_fd: RawFd,
+    ) -> anyhow::Result<Self> {
+        let fd = fd.as_raw_file_descriptor();
+
+        #[cfg(not(target_os = "linux"))]
+        return Self::non_atomic_dup2(fd, dest_fd);
+
+        let duped = unsafe { libc::dup3(fd, dest_fd, libc::O_CLOEXEC) };
```

Comment:
this line has to be enclosed in `#[cfg(target_os = "linux")]` block in order to compile+link on !linux systems

Reply:
Sorry not sure what you mean here. As is, this forces dup2 for non linux, otherwise, if linux, then it proceeds to attempt dup3 first. Is this not what you want here?

Reply:
Like I sketched out in the original comment; you need to enclose the block of code that calls dup3 in the cfg check, then under it, fall back to the non atomic dup2.

The issue to resolve is that you can only reference `dup3` on linux.  The program won't compile on any other platform unless dup3 is in a linux only block.

Reply:
The code above already does what you want here => if not linux then skip dup3 and do only dup2

Reply:
![image](https://user-images.githubusercontent.com/33977466/118161328-59df8380-b417-11eb-962f-33b74440597b.png)

Reply:
The intent of your change is correct, but the way it is expressed won't compile on mac or freebsd because `dup3` doesn't exist there.  You need to ensure that `dup3` is only referenced on linux, but right now it is referenced on all unix systems.

Reply:
Done

Reply:
GH actions need approval on each change, which is a bit of a drag, however, the freebsd CI is showing that this doesn't compile; it you structure it by enclosing the appropriate code in a block as I said in the original comment, you'll have the most success.

You can keep iterating against that until it goes green and then I can approve and run the rest of the platforms!

Reply:
@wez could you run the pipeline again please?